downsample_model.py

```diff
@@ -0,0 +1,242 @@
+    """
+    if len(splitting_dimensions) == 1:
+        splitting_size = splitting_dimensions[0]
+        splitted_tensor = torch.stack(weights.split(splitting_size, dim=-1))
+    elif len(splitting_dimensions) == 2:
+        splitting_size_x, splitting_size_y = splitting_dimensions
+        for i, splitting_size in enumerate(splitting_dimensions):
+            if i == 0:
+                splitted_tensor = torch.stack(weights.split(splitting_size, dim=0))
+            else:
+                splitted_tensor = torch.stack(splitted_tensor.split(splitting_size, dim=-1))
+        n_folds = splitted_tensor.shape[0] * splitted_tensor.shape[1]
+        splitted_tensor = splitted_tensor.contiguous().view(n_folds, splitting_size_x, splitting_size_y)
+
+    if aggregation_strategy == "mean":
+        return torch.mean(splitted_tensor, dim=0)
+    elif aggregation_strategy == "first":
+        return splitted_tensor[0]
+    elif aggregation_strategy == "last":
+        return splitted_tensor[-1]
+    else:
+        raise ValueError("Unknown aggregation strategy: {}".format(aggregation_strategy))
+
+def downsize_state_dict(state_dict, downsized_model_config, aggregation_strategy="mean"):
+    """
+        Take the old model's state dict and downsize it. The state dict should be already preprocessed
+        by the function `select_keys_from_state_dict` that will select only the relevant keys (relevant layers)
+        Args:
+            state_dict (`torch.state_dict`, *required*):
+                input state dict of the old model
+            downsized_model_config (`transformers.BloomConfig`, *required*):
+                config class of the shrinked model
+            aggregation_strategy (`str`, *required*):
+                aggregation strategy for the weights matrices. We first divide the weights matrix
+                into (1/(down_sampling_rate**2)) chunks for linear weights and (1/downsampling_rate)
+                for bias terms
+                - "first": Select only the first chunk
+                - "last": Select only the last chunk
+                - "mean": Average all the chunks 
+    """
+    downsized_state_dict = {}
+    for key in state_dict.keys():
+        splitting_dimensions = map_key_to_downsized_model(downsized_model_config.hidden_size, key)
+        weight_tensor = state_dict[key]
+        processed_weights = process_weights(weight_tensor, aggregation_strategy, splitting_dimensions)
+        downsized_state_dict[key] = processed_weights
+    return downsized_state_dict
+
+def main(args):
+    model_name = args.model_name
+    output_model_name = args.output_model_name
+    hidden_downsampling_rate = args.hidden_downsampling_rate
+    layer_downsampling_rate = args.layer_downsampling_rate
+    aggregation_strategy = args.aggregation_strategy
+    layer_selection_strategy = args.layer_selection_strategy
+    push_to_hub = args.push_to_hub
+    
+    config_old_model = BloomConfig.from_pretrained(model_name, use_auth_token=True)
```

Comment:
nit: just use `args.model_name`. Same for other args. Less code and less error-prone.

Reply:
Perfect! Fixed in 854dccf92fee29818a6cfc69e6ebc6e549a6ac7f